src/dune_rules/ml_sources.ml

```diff
@@ -424,7 +424,29 @@ let make dune_file ~dir ~scope ~lib_config ~loc ~lookup_vlib
           ~f:(fun acc ((dir : Path.Build.t), local, files) ->
             let modules = modules_of_files ~dialects ~dir ~files in
             let path = List.map local ~f:Module_name.of_string in
-            Module_trie.set_map acc path modules)
+            match Module_trie.set_map acc path modules with
+            | Ok s -> s
+            | Error module_ ->
+              let module_ =
+                match module_ with
+                | Leaf m ->
+                  Module.Source.files m |> List.hd |> Module.File.path
+                  |> Path.drop_optional_build_context
+                  |> Path.to_string_maybe_quoted
+                | Map _ -> assert false
+              in
+              let group =
+                dir |> Path.Build.drop_build_context_exn
+                |> Path.Source.to_string_maybe_quoted
+              in
+              User_error.raise ~loc
+                [ Pp.text
+                    "The following module and module group cannot exist in the \
+                     same executable or library exist because they correspond \
+                     to the same module path"
```

Comment:
```suggestion
                    "The following module and module group cannot coexist in the \
                     same executable or library because they correspond to the  \
                     same module path"
```